[CN] `cn` can't report rich errors in multiple unrelated functions

Consider the following code:
```c
int add1(int x)
/*@
requires x == 0i32;
ensures return == 2i32;
@*/
{
    return x + 1;
}

int add3(int x)
/*@
requires x == 0i32;
ensures return == 2i32;
@*/
{
    return x + 3;
}
```

`cn` correctly reports that `add3`'s spec doesn't hold, but doesn't say anything about `add1`, whose spec is also wrong:
```
> cn sample/add1.c
[1/2]: add3
sample/add1.c:16:5: error: Unprovable constraint
    return x + 3;
    ^~~~~~~~~~~~~ 
Constraint from sample/add1.c:13:9:
ensures return == 2i32;
        ^~~~~~~~~~~~~~~
```

 `cn --batch` will report that both functions "fail", but will not report why:
```
> cn --batch sample/add1.c
[1/2]: add3 -- fail
[2/2]: add1 -- fail
summary:
0 pass, 2 fail
```

It would be more user-friendly to be able to report as many errors as richly as possible - in this case, success would look like `cn` reporting both erroneous specs, with their locations, in a single invocation.

(I thought there might already be an issue for this behavior, but I couldn't find one. If there is one, I'm happy to deduplicate and direct conversation there.)

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -188,12 +188,35 @@ let with_well_formedness_check
     Printexc.raise_with_backtrace exc (Printexc.get_raw_backtrace ())
 
 
-let handle_type_error ~json ~state_file ~output_dir e =
+(** Report an error on [stderr] in an appropriate format: JSON if [json] is
+    true, or human-readable if not. *)
+let report_type_error
+  ~(json : bool)
+  ?(state_file : string option)
+  ?(output_dir : string option)
+  (error : TypeErrors.t)
+  : unit
+  =
   if json then
-    TypeErrors.report_json ?state_file ?output_dir e
+    TypeErrors.report_json ?state_file ?output_dir error
   else
-    TypeErrors.report_pretty ?state_file ?output_dir e;
-  match e.msg with TypeErrors.Unsupported _ -> exit 2 | _ -> exit 1
+    TypeErrors.report_pretty ?state_file ?output_dir error
+
+
+(** Exit with a code appropriate to the provided error. *)
+let exit_on_type_error (error : TypeErrors.t) : 'a =
+  match error.msg with TypeErrors.Unsupported _ -> exit 2 | _ -> exit 1
+
+
+(** Report the provided error, then exit. *)
+let handle_type_error
+  ~(json : bool)
+  ?(state_file : string option)
+  ?(output_dir : string option)
+  (error : TypeErrors.t)
+  =
+  report_type_error ~json ?state_file ?output_dir error;
+  exit_on_type_error error
 
 
 let well_formed
@@ -222,7 +245,7 @@ let well_formed
     ~use_peval
     ~no_inherit_loc
     ~magic_comment_char_dollar
-    ~handle_error:(handle_type_error ~json ~state_file ~output_dir)
+    ~handle_error:(handle_type_error ~json ?state_file ?output_dir)
     ~f:(fun ~prog5:_ ~ail_prog:_ ~statement_locs:_ ~paused:_ -> Resultat.return ())
 
 
@@ -303,7 +326,7 @@ let verify
     ~use_peval
     ~no_inherit_loc
     ~magic_comment_char_dollar (* Callbacks *)
-    ~handle_error:(handle_type_error ~json ~state_file ~output_dir)
+    ~handle_error:(handle_type_error ~json ?state_file ?output_dir)
     ~f:(fun ~prog5 ~ail_prog ~statement_locs ~paused ->
       match output_decorated with
       | None ->
@@ -524,8 +547,6 @@ module Verify_flags = struct
   let fail_fast =
     let doc = "Abort immediately after encountering a verification error" in
     Arg.(value & flag & info [ "fail-fast" ] ~doc)
-
-
 
 
   let slow_smt_threshold =

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -190,17 +190,13 @@ let with_well_formedness_check
 
 (** Report an error on [stderr] in an appropriate format: JSON if [json] is
     true, or human-readable if not. *)
-let report_type_error
-  ~(json : bool)
-  ?(state_file : string option)
-  ?(output_dir : string option)
-  (error : TypeErrors.t)
+let report_type_error ~(json : bool) ?(output_dir : string option) (error : TypeErrors.t)
   : unit
   =
   if json then
-    TypeErrors.report_json ?state_file ?output_dir error
+    TypeErrors.report_json ?output_dir error
   else
-    TypeErrors.report_pretty ?state_file ?output_dir error
+    TypeErrors.report_pretty ?output_dir error
 
 
 (** Exit with a code appropriate to the provided error. *)
@@ -209,13 +205,8 @@ let exit_on_type_error (error : TypeErrors.t) : 'a =
 
 
 (** Report the provided error, then exit. *)
-let handle_type_error
-  ~(json : bool)
-  ?(state_file : string option)
-  ?(output_dir : string option)
-  (error : TypeErrors.t)
-  =
-  report_type_error ~json ?state_file ?output_dir error;
+let handle_type_error ~(json : bool) ?(output_dir : string option) (error : TypeErrors.t) =
+  report_type_error ~json ?output_dir error;
   exit_on_type_error error
 
 
@@ -225,7 +216,6 @@ let well_formed
   incl_dirs
   incl_files
   json
-  state_file
   output_dir
   csv_times
   log_times
@@ -245,7 +235,7 @@ let well_formed
     ~use_peval
     ~no_inherit_loc
     ~magic_comment_char_dollar
-    ~handle_error:(handle_type_error ~json ?state_file ?output_dir)
+    ~handle_error:(handle_type_error ~json ?output_dir)
     ~f:(fun ~prog5:_ ~ail_prog:_ ~statement_locs:_ ~paused:_ -> Resultat.return ())
 
 
@@ -262,7 +252,6 @@ let verify
   slow_smt_dir
   no_timestamps
   json
-  state_file
   output_dir
   diag
   lemmata
@@ -327,7 +316,7 @@ let verify
     ~use_peval
     ~no_inherit_loc
     ~magic_comment_char_dollar (* Callbacks *)
-    ~handle_error:(handle_type_error ~json ?state_file ?output_dir)
+    ~handle_error:(handle_type_error ~json ?output_dir)
     ~f:(fun ~prog5 ~ail_prog ~statement_locs ~paused ->
       match output_decorated with
       | None ->
@@ -335,7 +324,7 @@ let verify
           let open Typing in
           let@ errors = Check.time_check_c_functions functions in
           if not quiet then
-            List.iter (report_type_error ~json ?state_file ?output_dir) errors;
+            List.iter (report_type_error ~json ?output_dir) errors;
           Check.generate_lemmas lemmas lemmata
         in
         Typing.run_from_pause check paused
@@ -570,11 +559,6 @@ module Verify_flags = struct
     Arg.(value & opt (some string) None & info [ "slow-smt-dir" ] ~docv:"FILE" ~doc)
 
 
-  let state_file =
-    let doc = "file in which to output the state" in
-    Arg.(value & opt (some string) None & info [ "state-file" ] ~docv:"FILE" ~doc)
-
-
   let diag =
     let doc = "explore branching diagnostics with key string" in
     Arg.(value & opt (some string) None & info [ "diag" ] ~doc)
@@ -634,7 +618,7 @@ module Verify_flags = struct
 
 
   let output_dir =
-    let doc = "directory in which to output state files (overridden by --state-file)" in
+    let doc = "directory in which to output state files" in
     Arg.(value & opt (some string) None & info [ "output-dir" ] ~docv:"FILE" ~doc)
 end
 
@@ -691,7 +675,6 @@ let wf_cmd =
     $ Common_flags.incl_dirs
     $ Common_flags.incl_files
     $ Verify_flags.json
-    $ Verify_flags.state_file
     $ Verify_flags.output_dir
     $ Common_flags.csv_times
     $ Common_flags.log_times
@@ -728,7 +711,6 @@ let verify_t : unit Term.t =
   $ Verify_flags.slow_smt_dir
   $ Common_flags.no_timestamps
   $ Verify_flags.json
-  $ Verify_flags.state_file
   $ Verify_flags.output_dir
   $ Verify_flags.diag
   $ Lemma_flags.lemmata

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -199,15 +199,28 @@ let report_type_error ~(json : bool) ?(output_dir : string option) (error : Type
     TypeErrors.report_pretty ?output_dir error
 
 
-(** Exit with a code appropriate to the provided error. *)
-let exit_on_type_error (error : TypeErrors.t) : 'a =
-  match error.msg with TypeErrors.Unsupported _ -> exit 2 | _ -> exit 1
+(** Generate an appropriate exit code for the provided error. *)
+let exit_code_of_error (error : TypeErrors.t) : int =
+  match error.msg with TypeErrors.Unsupported _ -> 2 | _ -> 1
+
+
+(** In the presence of nonempty errors, generate an appropriate exit code for
+    them. *)
+let exit_code_of_errors (errors : TypeErrors.t list) : int option =
+  (* We arbitrarily choose to make any `Unsupported` message dominate all other
+     messages. *)
+  let is_unsupported (err : TypeErrors.t) =
+    match err.msg with TypeErrors.Unsupported _ -> true | _ -> false
+  in
+  match errors with
+  | [] -> None
+  | _ -> Some (if List.exists is_unsupported errors then 2 else 1)
 
 
 (** Report the provided error, then exit. *)
 let handle_type_error ~(json : bool) ?(output_dir : string option) (error : TypeErrors.t) =
   report_type_error ~json ?output_dir error;
-  exit_on_type_error error
+  exit (exit_code_of_error error)
 
 
 let well_formed
@@ -325,6 +338,7 @@ let verify
           let@ errors = Check.time_check_c_functions functions in
           if not quiet then
             List.iter (report_type_error ~json ?output_dir) errors;
+          Option.fold ~none:() ~some:exit (exit_code_of_errors errors);
           Check.generate_lemmas lemmas lemmata
         in
         Typing.run_from_pause check paused

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -306,7 +306,13 @@ let verify
     ~handle_error:(handle_type_error ~json ~state_file ~output_dir)
     ~f:(fun ~prog5 ~ail_prog ~statement_locs ~paused ->
       match output_decorated with
-      | None -> Typing.run_from_pause (fun paused -> Check.check paused lemmata) paused
+      | None ->
+        let check (functions, lemmas) =
+          let open Typing in
+          let@ _errors = Check.time_check_c_functions functions in
+          Check.generate_lemmas lemmas lemmata
+        in
+        Typing.run_from_pause check paused
       | Some output_filename ->
         Cerb_colour.without_colour
           (fun () ->

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -261,7 +261,7 @@ let verify
   use_vip
   no_use_ity
   use_peval
-  batch
+  fail_fast
   no_inherit_loc
   magic_comment_char_dollar
   =
@@ -288,7 +288,7 @@ let verify
   Solver.solver_flags := solver_flags;
   Check.skip_and_only := (opt_comma_split skip, opt_comma_split only);
   IndexTerms.use_vip := use_vip;
-  Check.batch := batch;
+  Check.fail_fast := fail_fast;
   Diagnostics.diag_string := diag;
   WellTyped.use_ity := not no_use_ity;
   Sym.executable_spec_enabled := Option.is_some output_decorated;
@@ -515,12 +515,11 @@ module Verify_flags = struct
       & info [ "locs" ] ~docv:"HEX" ~doc)
 
 
-  let batch =
-    let doc =
-      "Type check functions in batch/do not stop on first type error (unless `only` is \
-       used)"
-    in
-    Arg.(value & flag & info [ "batch" ] ~doc)
+  let fail_fast =
+    let doc = "Abort immediately after encountering a verification error" in
+    Arg.(value & flag & info [ "fail-fast" ] ~doc)
+
+
 
 
   let slow_smt_threshold =
@@ -716,7 +715,7 @@ let verify_t : unit Term.t =
   $ Verify_flags.use_vip
   $ Common_flags.no_use_ity
   $ Common_flags.use_peval
-  $ Verify_flags.batch
+  $ Verify_flags.fail_fast
   $ Common_flags.no_inherit_loc
   $ Common_flags.magic_comment_char_dollar
 

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -332,7 +332,8 @@ let verify
       | None ->
         let check (functions, lemmas) =
           let open Typing in
-          let@ _errors = Check.time_check_c_functions functions in
+          let@ errors = Check.time_check_c_functions functions in
+          List.iter (report_type_error ~json ?state_file ?output_dir) errors;
           Check.generate_lemmas lemmas lemmata
         in
         Typing.run_from_pause check paused

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -285,6 +285,7 @@ let verify
   no_use_ity
   use_peval
   fail_fast
+  quiet
   no_inherit_loc
   magic_comment_char_dollar
   =
@@ -333,7 +334,8 @@ let verify
         let check (functions, lemmas) =
           let open Typing in
           let@ errors = Check.time_check_c_functions functions in
-          List.iter (report_type_error ~json ?state_file ?output_dir) errors;
+          if not quiet then
+            List.iter (report_type_error ~json ?state_file ?output_dir) errors;
           Check.generate_lemmas lemmas lemmata
         in
         Typing.run_from_pause check paused
@@ -550,6 +552,11 @@ module Verify_flags = struct
     Arg.(value & flag & info [ "fail-fast" ] ~doc)
 
 
+  let quiet =
+    let doc = "Only report success and failure, rather than rich errors" in
+    Arg.(value & flag & info [ "quiet" ] ~doc)
+
+
   let slow_smt_threshold =
     let doc = "Set the time threshold (in seconds) for logging slow smt queries." in
     Arg.(value & opt (some float) None & info [ "slow-smt" ] ~docv:"TIMEOUT" ~doc)
@@ -744,6 +751,7 @@ let verify_t : unit Term.t =
   $ Common_flags.no_use_ity
   $ Common_flags.use_peval
   $ Verify_flags.fail_fast
+  $ Verify_flags.quiet
   $ Common_flags.no_inherit_loc
   $ Common_flags.magic_comment_char_dollar
 

--- a/backend/cn/bin/main.ml
+++ b/backend/cn/bin/main.ml
@@ -192,7 +192,7 @@ let handle_type_error ~json ~state_file ~output_dir e =
   if json then
     TypeErrors.report_json ?state_file ?output_dir e
   else
-    TypeErrors.report ?state_file ?output_dir e;
+    TypeErrors.report_pretty ?state_file ?output_dir e;
   match e.msg with TypeErrors.Unsupported _ -> exit 2 | _ -> exit 1
 
 

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2077,27 +2077,32 @@ let check_c_functions_fast (funs : c_function list) : unit m =
   return ()
 
 
+(** Check the provided C functions, each in an isolated context, capturing and
+    counting failures. The result represents [(num_passed, num_failed)]. *)
+let check_c_functions_all (funs : c_function list) : (int * int) m =
+  let total = List.length funs in
+  let check_and_count (num_checked, pass, fail) c_fn =
+    let fn_name = c_function_name c_fn in
+    let@ outcome = sandbox (check_c_function c_fn) in
+    let checked = num_checked + 1 in
+    match outcome with
+    | Ok () ->
+      progress_simple (of_total checked total) (fn_name ^ " -- pass");
+      return (checked, pass + 1, fail)
+    | Error _ ->
+      progress_simple (of_total checked total) (fn_name ^ " -- fail");
+      return (checked, pass, fail + 1)
+  in
+  let@ _, pass, fail = ListM.fold_leftM check_and_count (0, 0, 0) funs in
+  return (pass, fail)
+
+
 let check_c_functions funs =
   let selected_funs = select_functions funs in
-  let number_entries = List.length selected_funs in
   match !batch with
   | false -> check_c_functions_fast selected_funs
   | true ->
-    let@ _, pass, fail =
-      ListM.fold_leftM
-        (fun (counter, pass, fail) c_fn ->
-          let fn_name = c_function_name c_fn in
-          let@ outcome = sandbox (check_c_function c_fn) in
-          match outcome with
-          | Ok () ->
-            progress_simple (of_total (counter + 1) number_entries) (fn_name ^ " -- pass");
-            return (counter + 1, pass + 1, fail)
-          | Error _err ->
-            progress_simple (of_total (counter + 1) number_entries) (fn_name ^ " -- fail");
-            return (counter + 1, pass, fail + 1))
-        (0, 0, 0)
-        selected_funs
-    in
+    let@ pass, fail = check_c_functions_all selected_funs in
     print
       stdout
       (item "summary" (int pass ^^^ !^"pass" ^^ comma ^^^ int fail ^^^ !^"fail"));

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2101,7 +2101,9 @@ let check_c_functions_all (funs : c_function list) : TypeErrors.t list m =
   return (List.rev errors)
 
 
-let check_c_functions funs =
+(** Downselect from the provided functions with [select_functions] and check the
+    results. When [batch] is set, report summary statistics of all errors. *)
+let check_c_functions (funs : c_function list) : unit m =
   let selected_funs = select_functions funs in
   match !batch with
   | false -> check_c_functions_fast selected_funs

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2070,29 +2070,26 @@ let check_c_functions funs =
   | false ->
     let@ _ =
       ListM.mapiM
-        (fun counter (fsym, (loc, args_and_body)) ->
+        (fun counter c_fn ->
           let () =
-            progress_simple (of_total (counter + 1) number_entries) (Sym.pp_string fsym)
+            progress_simple (of_total (counter + 1) number_entries) (c_function_name c_fn)
           in
-          check_procedure loc fsym args_and_body)
+          check_c_function c_fn)
         selected_funs
     in
     return ()
   | true ->
     let@ _, pass, fail =
       ListM.fold_leftM
-        (fun (counter, pass, fail) (fsym, (loc, args_and_body)) ->
-          let@ outcome = sandbox (check_procedure loc fsym args_and_body) in
+        (fun (counter, pass, fail) c_fn ->
+          let fn_name = c_function_name c_fn in
+          let@ outcome = sandbox (check_c_function c_fn) in
           match outcome with
-          | Ok _ ->
-            progress_simple
-              (of_total (counter + 1) number_entries)
-              (Sym.pp_string fsym ^ " -- pass");
+          | Ok () ->
+            progress_simple (of_total (counter + 1) number_entries) (fn_name ^ " -- pass");
             return (counter + 1, pass + 1, fail)
-          | Error _ ->
-            progress_simple
-              (of_total (counter + 1) number_entries)
-              (Sym.pp_string fsym ^ " -- fail");
+          | Error _err ->
+            progress_simple (of_total (counter + 1) number_entries) (fn_name ^ " -- fail");
             return (counter + 1, pass, fail + 1))
         (0, 0, 0)
         selected_funs

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2030,7 +2030,11 @@ let wf_check_and_record_functions mu_funs mu_call_sigs =
     ([], [])
 
 
-let check_c_functions funs =
+type c_function = symbol * (loc * basetype mu_proc_args_and_body)
+
+(** Filter functions according to [skip_and_only]: first according to "only",
+    then according to "skip" *)
+let select_functions (funs : c_function list) : c_function list =
   let matches_str s fsym = String.equal s (Sym.pp_string fsym) in
   let str_fsyms s =
     match List.filter (matches_str s) (List.map fst funs) with
@@ -2047,9 +2051,11 @@ let check_c_functions funs =
     | [] -> funs
     | _ss -> List.filter (fun (fsym, _) -> SymSet.mem fsym only) funs
   in
-  let selected_funs =
-    List.filter (fun (fsym, _) -> not (SymSet.mem fsym skip)) only_funs
-  in
+  List.filter (fun (fsym, _) -> not (SymSet.mem fsym skip)) only_funs
+
+
+let check_c_functions funs =
+  let selected_funs = select_functions funs in
   let number_entries = List.length selected_funs in
   match !batch with
   | false ->

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2063,21 +2063,25 @@ let check_c_function ((fsym, (loc, args_and_body)) : c_function) : unit m =
   check_procedure loc fsym args_and_body
 
 
+(** Check the provided C functions. Failure of any check will short-circuit the
+    remainder of the checks. *)
+let check_c_functions_fast (funs : c_function list) : unit m =
+  let total = List.length funs in
+  let@ _ =
+    ListM.mapiM
+      (fun counter c_fn ->
+        let () = progress_simple (of_total (counter + 1) total) (c_function_name c_fn) in
+        check_c_function c_fn)
+      funs
+  in
+  return ()
+
+
 let check_c_functions funs =
   let selected_funs = select_functions funs in
   let number_entries = List.length selected_funs in
   match !batch with
-  | false ->
-    let@ _ =
-      ListM.mapiM
-        (fun counter c_fn ->
-          let () =
-            progress_simple (of_total (counter + 1) number_entries) (c_function_name c_fn)
-          in
-          check_c_function c_fn)
-        selected_funs
-    in
-    return ()
+  | false -> check_c_functions_fast selected_funs
   | true ->
     let@ _, pass, fail =
       ListM.fold_leftM

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2063,18 +2063,28 @@ let check_c_function ((fsym, (loc, args_and_body)) : c_function) : unit m =
   check_procedure loc fsym args_and_body
 
 
-(** Check the provided C functions. Failure of any check will short-circuit the
-    remainder of the checks. *)
-let check_c_functions_fast (funs : c_function list) : unit m =
+(** Check the provided C functions. The first failed check will short-circuit
+    the remainder of the checks, and the associated error will be returned as
+    [Some]. *)
+let check_c_functions_fast (funs : c_function list) : TypeErrors.t option m =
   let total = List.length funs in
-  let@ _ =
-    ListM.mapiM
-      (fun counter c_fn ->
-        let () = progress_simple (of_total (counter + 1) total) (c_function_name c_fn) in
-        check_c_function c_fn)
-      funs
+  let check_and_record (num_checked, prev_error) c_fn =
+    match prev_error with
+    | Some _ -> return (num_checked, prev_error)
+    | None ->
+      let fn_name = c_function_name c_fn in
+      let@ outcome = sandbox (check_c_function c_fn) in
+      let checked = num_checked + 1 in
+      (match outcome with
+       | Ok () ->
+         progress_simple (of_total checked total) (fn_name ^ " -- pass");
+         return (checked, None)
+       | Error err ->
+         progress_simple (of_total checked total) (fn_name ^ " -- fail");
+         return (checked, Some err))
   in
-  return ()
+  let@ _num_checked, error = ListM.fold_leftM check_and_record (0, None) funs in
+  return error
 
 
 (** Check the provided C functions, each in an isolated context, capturing any
@@ -2106,7 +2116,9 @@ let check_c_functions_all (funs : c_function list) : TypeErrors.t list m =
 let check_c_functions (funs : c_function list) : unit m =
   let selected_funs = select_functions funs in
   match !batch with
-  | false -> check_c_functions_fast selected_funs
+  | false ->
+    let@ _ = check_c_functions_fast selected_funs in
+    return ()
   | true ->
     let@ errors = check_c_functions_all selected_funs in
     let fail = List.length errors in

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2077,24 +2077,28 @@ let check_c_functions_fast (funs : c_function list) : unit m =
   return ()
 
 
-(** Check the provided C functions, each in an isolated context, capturing and
-    counting failures. The result represents [(num_passed, num_failed)]. *)
-let check_c_functions_all (funs : c_function list) : (int * int) m =
+(** Check the provided C functions, each in an isolated context, capturing any
+    (monadic) check failures and returning them. All checks will be performed
+    regardless of intermediate failures. The result's order is determined by
+    the input's order: if function [f] appears before function [g], then
+    function [f]'s error (if any) will appear before function [g]'s error (if
+    any). *)
+let check_c_functions_all (funs : c_function list) : TypeErrors.t list m =
   let total = List.length funs in
-  let check_and_count (num_checked, pass, fail) c_fn =
+  let check_and_record (num_checked, errors) c_fn =
     let fn_name = c_function_name c_fn in
     let@ outcome = sandbox (check_c_function c_fn) in
     let checked = num_checked + 1 in
     match outcome with
     | Ok () ->
       progress_simple (of_total checked total) (fn_name ^ " -- pass");
-      return (checked, pass + 1, fail)
-    | Error _ ->
+      return (checked, errors)
+    | Error err ->
       progress_simple (of_total checked total) (fn_name ^ " -- fail");
-      return (checked, pass, fail + 1)
+      return (checked, err :: errors)
   in
-  let@ _, pass, fail = ListM.fold_leftM check_and_count (0, 0, 0) funs in
-  return (pass, fail)
+  let@ _num_checked, errors = ListM.fold_leftM check_and_record (0, []) funs in
+  return (List.rev errors)
 
 
 let check_c_functions funs =
@@ -2102,7 +2106,9 @@ let check_c_functions funs =
   match !batch with
   | false -> check_c_functions_fast selected_funs
   | true ->
-    let@ pass, fail = check_c_functions_all selected_funs in
+    let@ errors = check_c_functions_all selected_funs in
+    let fail = List.length errors in
+    let pass = List.length selected_funs - fail in
     print
       stdout
       (item "summary" (int pass ^^^ !^"pass" ^^ comma ^^^ int fail ^^^ !^"fail"));

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2261,15 +2261,20 @@ let check_decls_lemmata_fun_specs (mu_file : unit mu_file) =
   return (List.rev checked, lemmata)
 
 
-let check (checked, lemmata) o_lemma_mode =
+(** With CSV timing enabled, check the provided functions with
+    [check_c_functions]. See that function for more information on the
+    semantics of checking. *)
+let time_check_c_functions (checked : c_function list) : TypeErrors.t list m =
   Cerb_debug.begin_csv_timing () (*type checking functions*);
-  let@ _ = check_c_functions checked in
+  let@ errors = check_c_functions checked in
   Cerb_debug.end_csv_timing "type checking functions";
+  return errors
+
+
+let generate_lemmas lemmata o_lemma_mode =
   let@ global = get_global () in
   match o_lemma_mode with
-  | Some mode ->
-    let@ _ = embed_resultat (Lemmata.generate global mode lemmata) in
-    return ()
+  | Some mode -> embed_resultat (Lemmata.generate global mode lemmata)
   | None -> return ()
 
 (* TODO:

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2032,6 +2032,10 @@ let wf_check_and_record_functions mu_funs mu_call_sigs =
 
 type c_function = symbol * (loc * basetype mu_proc_args_and_body)
 
+let c_function_name ((fsym, (_loc, _args_and_body)) : c_function) : string =
+  Sym.pp_string fsym
+
+
 (** Filter functions according to [skip_and_only]: first according to "only",
     then according to "skip" *)
 let select_functions (funs : c_function list) : c_function list =
@@ -2052,6 +2056,11 @@ let select_functions (funs : c_function list) : c_function list =
     | _ss -> List.filter (fun (fsym, _) -> SymSet.mem fsym only) funs
   in
   List.filter (fun (fsym, _) -> not (SymSet.mem fsym skip)) only_funs
+
+
+(** Check a single C function. Failure of the check is encoded monadically. *)
+let check_c_function ((fsym, (loc, args_and_body)) : c_function) : unit m =
+  check_procedure loc fsym args_and_body
 
 
 let check_c_functions funs =

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -2250,7 +2250,7 @@ let check_decls_lemmata_fun_specs (mu_file : unit mu_file) =
   in
   Pp.debug 3 (lazy (Pp.headline "type-checked C functions and specifications."));
   Cerb_debug.end_csv_timing "decl, lemmata, function specification checking";
-  return (checked, lemmata)
+  return (List.rev checked, lemmata)
 
 
 let check (checked, lemmata) o_lemma_mode =

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -1863,7 +1863,9 @@ let check_procedure (loc : loc) (fsym : Sym.t) (args_and_body : _ mu_proc_args_a
 
 let skip_and_only = ref (([] : string list), ([] : string list))
 
-let batch = ref false
+(** When set, causes verification of multiple functions to abort as soon as a
+    single function fails to verify. *)
+let fail_fast = ref false
 
 let record_tagdefs tagDefs =
   PmapM.iterM
@@ -2112,21 +2114,15 @@ let check_c_functions_all (funs : c_function list) : TypeErrors.t list m =
 
 
 (** Downselect from the provided functions with [select_functions] and check the
-    results. When [batch] is set, report summary statistics of all errors. *)
-let check_c_functions (funs : c_function list) : unit m =
+    results. Errors in checking are captured, collected, and returned. When
+    [fail_fast] is set, the first error encountered will halt checking. *)
+let check_c_functions (funs : c_function list) : TypeErrors.t list m =
   let selected_funs = select_functions funs in
-  match !batch with
-  | false ->
-    let@ _ = check_c_functions_fast selected_funs in
-    return ()
+  match !fail_fast with
   | true ->
-    let@ errors = check_c_functions_all selected_funs in
-    let fail = List.length errors in
-    let pass = List.length selected_funs - fail in
-    print
-      stdout
-      (item "summary" (int pass ^^^ !^"pass" ^^ comma ^^^ int fail ^^^ !^"fail"));
-    return ()
+    let@ error_opt = check_c_functions_fast selected_funs in
+    return (Option.to_list error_opt)
+  | false -> check_c_functions_all selected_funs
 
 
 (* (Sym.t * (Locations.t * ArgumentTypes.lemmat)) list *)
@@ -2267,7 +2263,7 @@ let check_decls_lemmata_fun_specs (mu_file : unit mu_file) =
 
 let check (checked, lemmata) o_lemma_mode =
   Cerb_debug.begin_csv_timing () (*type checking functions*);
-  let@ () = check_c_functions checked in
+  let@ _ = check_c_functions checked in
   Cerb_debug.end_csv_timing "type checking functions";
   let@ global = get_global () in
   match o_lemma_mode with

--- a/backend/cn/lib/typeErrors.ml
+++ b/backend/cn/lib/typeErrors.ml
@@ -549,11 +549,16 @@ let canonicalize (path : string) : string =
     path
 
 
-(** Create a .html file, with a random name, in [output_dir], which must already
-    exist. If no directory is provided, the file is created in the system
-    temporary directory instead. *)
-let mk_state_file_name ~(output_dir : string option) : string =
-  let prefix = "state_" in
+(** Create a filename derived from the given error location, and create a file
+    with that name in [output_dir], which must already exist. If no directory
+    is provided, the file is created in the system temporary directory
+    instead. *)
+let mk_state_file_name ?(output_dir : string option) (loc : Cerb_location.t) : string =
+  let prefix =
+    match Cerb_location.get_filename loc with
+    | None -> "state_"
+    | Some filename -> "state_" ^ Filename.basename filename ^ "_"
+  in
   let canonical_output_dir = Option.map canonicalize output_dir in
   try Filename.temp_file ?temp_dir:canonical_output_dir prefix ".html" with
   | Sys_error _ ->
@@ -575,7 +580,9 @@ let report_pretty ?state_file:to_ ?output_dir:dir_ { loc; msg } =
     | Some state ->
       (* Decide where to write the state *)
       let state_error_file =
-        match to_ with Some file -> file | None -> mk_state_file_name ~output_dir:dir_
+        match to_ with
+        | Some file -> file
+        | None -> mk_state_file_name ?output_dir:dir_ loc
       in
       let link = Report.make state_error_file (Cerb_location.get_filename loc) state in
       let msg = !^"State file:" ^^^ !^("file://" ^ link) in
@@ -593,7 +600,9 @@ let report_json ?state_file:to_ ?output_dir:dir_ { loc; msg } =
     | Some state ->
       (* Decide where to write the state *)
       let file =
-        match to_ with Some file -> file | None -> mk_state_file_name ~output_dir:dir_
+        match to_ with
+        | Some file -> file
+        | None -> mk_state_file_name ?output_dir:dir_ loc
       in
       let link = Report.make file (Cerb_location.get_filename loc) state in
       `String link

--- a/backend/cn/lib/typeErrors.ml
+++ b/backend/cn/lib/typeErrors.ml
@@ -572,18 +572,13 @@ let mk_state_file_name ?(output_dir : string option) (loc : Cerb_location.t) : s
     error contains enough information to create an HTML state report, generate
     one in [output_dir] (or, failing that, the system temporary directory) and
     print a link to it. *)
-let report_pretty ?state_file:to_ ?output_dir:dir_ { loc; msg } =
+let report_pretty ?output_dir:dir_ { loc; msg } =
   (* stealing some logic from pp_errors *)
   let report = pp_message msg in
   let consider =
     match report.state with
     | Some state ->
-      (* Decide where to write the state *)
-      let state_error_file =
-        match to_ with
-        | Some file -> file
-        | None -> mk_state_file_name ?output_dir:dir_ loc
-      in
+      let state_error_file = mk_state_file_name ?output_dir:dir_ loc in
       let link = Report.make state_error_file (Cerb_location.get_filename loc) state in
       let msg = !^"State file:" ^^^ !^("file://" ^ link) in
       Some msg
@@ -593,17 +588,12 @@ let report_pretty ?state_file:to_ ?output_dir:dir_ { loc; msg } =
 
 
 (* stealing some logic from pp_errors *)
-let report_json ?state_file:to_ ?output_dir:dir_ { loc; msg } =
+let report_json ?output_dir:dir_ { loc; msg } =
   let report = pp_message msg in
   let state_error_file =
     match report.state with
     | Some state ->
-      (* Decide where to write the state *)
-      let file =
-        match to_ with
-        | Some file -> file
-        | None -> mk_state_file_name ?output_dir:dir_ loc
-      in
+      let file = mk_state_file_name ?output_dir:dir_ loc in
       let link = Report.make file (Cerb_location.get_filename loc) state in
       `String link
     | None -> `Null

--- a/backend/cn/lib/typeErrors.ml
+++ b/backend/cn/lib/typeErrors.ml
@@ -540,36 +540,44 @@ let pp_message te =
 
 type t = type_error
 
-(* stealing some logic from pp_errors *)
-let report ?state_file:to_ ?output_dir:dir_ { loc; msg } =
+(** Convert a possibly-relative filepath into an absolute one. *)
+let canonicalize (path : string) : string =
+  if Filename.is_relative path then (
+    let current_dir = Sys.getcwd () in
+    Filename.concat current_dir path)
+  else
+    path
+
+
+(** Create a .html file, with a random name, in [output_dir], which must already
+    exist. If no directory is provided, the file is created in the system
+    temporary directory instead. *)
+let mk_state_file_name ~(output_dir : string option) : string =
+  let prefix = "state_" in
+  let canonical_output_dir = Option.map canonicalize output_dir in
+  try Filename.temp_file ?temp_dir:canonical_output_dir prefix ".html" with
+  | Sys_error _ ->
+    Printf.eprintf
+      "Warning: failed to create state file in the specified directory. Using the \
+       default output directory instead.\n";
+    Filename.temp_file prefix ".html"
+
+
+(** Format the error for human readability and print it to [stderr]. if the
+    error contains enough information to create an HTML state report, generate
+    one in [output_dir] (or, failing that, the system temporary directory) and
+    print a link to it. *)
+let report_pretty ?state_file:to_ ?output_dir:dir_ { loc; msg } =
+  (* stealing some logic from pp_errors *)
   let report = pp_message msg in
   let consider =
     match report.state with
     | Some state ->
       (* Decide where to write the state *)
       let state_error_file =
-        match to_ with
-        | Some file -> file
-        | None ->
-          (match dir_ with
-           | Some dir ->
-             let full_output_dir =
-               if Filename.is_relative dir then (
-                 let current_dir = Sys.getcwd () in
-                 Filename.concat current_dir dir)
-               else
-                 dir
-             in
-             (try Filename.temp_file ~temp_dir:full_output_dir "state_" ".html" with
-              | Sys_error _ ->
-                Printf.eprintf
-                  "Warning: failed to create state file in the specified directory. \
-                   Using the default output directory instead.\n";
-                Filename.temp_file "state_" ".html")
-           | None -> Filename.temp_file "state_" ".html")
+        match to_ with Some file -> file | None -> mk_state_file_name ~output_dir:dir_
       in
       let link = Report.make state_error_file (Cerb_location.get_filename loc) state in
-      (* let link = Report.make state_error_file state in *)
       let msg = !^"State file:" ^^^ !^("file://" ^ link) in
       Some msg
     | None -> None
@@ -585,25 +593,7 @@ let report_json ?state_file:to_ ?output_dir:dir_ { loc; msg } =
     | Some state ->
       (* Decide where to write the state *)
       let file =
-        match to_ with
-        | Some file -> file
-        | None ->
-          (match dir_ with
-           | Some dir ->
-             let full_output_dir =
-               if Filename.is_relative dir then (
-                 let current_dir = Sys.getcwd () in
-                 Filename.concat current_dir dir)
-               else
-                 dir
-             in
-             (try Filename.temp_file ~temp_dir:full_output_dir "" ".cn-state" with
-              | Sys_error _ ->
-                Printf.eprintf
-                  "Warning: failed to create state file in the specified directory. \
-                   Using the default output directory instead.\n";
-                Filename.temp_file "" ".cn-state")
-           | None -> Filename.temp_file "" ".cn-state")
+        match to_ with Some file -> file | None -> mk_state_file_name ~output_dir:dir_
       in
       let link = Report.make file (Cerb_location.get_filename loc) state in
       `String link

